### PR TITLE
Prefer bitsize variant over standard integral types

### LINKED ISSUE #69: Print integral types as bit sizes instead of standard types
Hello !

Another idea for improvements:

image

As for now, integral types are shown as `int`, `unsigned int`, etc

An idea is to print them as `uint8`, `int8`, `uint16`, etc

This is because an `int` does represent the size of the integral only for the platform it currently run.

I think a lot of developers do not use standard types (`int`, `char`, ...) but typedefs types (`uint32_t`, `uint8_t`, ...) just like me!

I think this is just a matter of updating theses lines with proper types even if there could be a better more generic way:
image

If you agree to this feature and you are not available, it will be my pleasure to provide a PR!

### PR DESCRIPTION
**Pull request by nocoinman**
Fixes #69 

Have adapted some tests, replaced some types in const, pointer tests to keep code clean.

**Note** : Although the CI checks are all passing, the standard does not mandate fixed width integer types to be available. Though very rare, there might be cases when these are undefined (embedded applications is what I can think of). I am not sure if if this project in intended to support such applications. If that is the case we could wrap each association with a feature test macro. Something like

```cpp
#if defined(INT16_MAX)
DBG_MACRO_REGISTER_TYPE_ASSOC(short, int16_t)
#else 
DBG_MACRO_REGISTER_TYPE_ASSOC(short, short)
#endif 
```
Or if you have any other elegant suggestions please let me know!

### CODE REVIEW

File: tests/example.cpp [hunk 1]
```diff
@@ -18,9 +18,11 @@ int main() {
   const int b = dbg(3 * a) + 1;  // [example.cpp:18 (main)] 3 * a = 6 (int)
 
   std::vector<int> numbers{b, 13, 42};
-  dbg(numbers);  // [example.cpp:21 (main)] numbers = {7, 13, 42} (std::vector<int>)
+  dbg(numbers);  // [example.cpp:21 (main)] numbers = {7, 13, 42}
+                 // (std::vector<int>)
```
Comment (sharkdp) [inline tests/example.cpp:22]: Can we please revert the formatting here (and maybe replace `int` with `int32_t`? both in the output/comment as well as in the code.

Comment (sharkdp) [reply tests/example.cpp:22]: Similar below.

We should probably user `int32_t` throughout this `example.cpp`?

Also, once this file is updated, can you please copy it over to the README example?

**Review by sharkdp — COMMENTED** [review]
Thank you very much!

**Comment by sharkdp** [comment]
> **Note** : Although the CI checks are all passing, the standard does not mandate fixed width integer types to be available. Though very rare, there might be cases when these are undefined (embedded applications is what I can think of). I am not sure if if this project in intended to support such applications.

Thank you for thinking about this. I think it's fine to rely on the availability. For now, we do support the platforms and compilers that are tested in CI (GCC, Clang, MSVC). If someone wants to add support for additional platforms or compilers, we can re-evaluate this.

**Comment by nocoinman** [comment]
Another behaviour to note, even though the associations are registered, type names deduced from `__PRETTY_FUNCTION__` and friends will still show `int` instead of `int32_t`. See below

```cpp
int var{};
dbg(var);  // var = 0 (int32_t)

int arr1[3]{};
dbg(arr1);  // arr1 = {0, 0, 0} (int [3])

std::array<int, 3> arr2{};
dbg(arr2);  // arr2 = {0, 0, 0} (std::array<int, 3>)

std::array<std::pair<int, int>, 3> arr3{};
dbg(arr3);  // arr3 = {{0, 0}, {0, 0}, {0, 0}} (std::array<std::pair<int, int>, 3>)
dbg(arr3[0]);  // arr3[0] = {0, 0} (std::pair<int32_t, int32_t>&)
```

I am not sure how to fix this apart from specialising `get_type_name` for such common cases. What are your thoughts on this?

File: README.md [hunk 2]
```diff
@@ -46,7 +47,7 @@ int main() {
 
 The code above produces this output ([try it yourself](https://repl.it/@sharkdp/dbg-macro-demo)):
 
-![dbg(…) macro output](https://i.imgur.com/NHEYk9A.png)
+![dbg(…) macro output](https://i.imgur.com/NVTzGRk.png)
```
Comment (sharkdp) [inline README.md:50]: :heart:

**Comment by sharkdp** [comment]
Thank you. This looks (almost) good to go. Can you please remove the `tests/example` binary?

> I am not sure how to fix this apart from specialising get_type_name for such common cases. What are your thoughts on this?

Hm. Adding a specialization for C-style arrays and `std::array` would probably be a good idea, to avoid inconsistencies/confusion. But feel free to skip this for now.

**Review by sharkdp — APPROVED** [review]
Thank you for the update!

**Comment by nocoinman** [comment]
Love this project! Taught me a lot :smile:

**Pull request merged**